lib/core/src/Cardano/Wallet.hs

```diff
@@ -2087,6 +2100,41 @@ guardHardIndex ix =
     then throwE $ ErrIndexOutOfBound minBound maxBound ix
     else pure (Index $ getDerivationIndex ix)
 
+updateCosigner
+    :: forall ctx s k n.
+        ( HasDBLayer s k ctx
+        , s ~ SharedState n k
+        , MkKeyFingerprint k (Proxy n, k 'AddressK CC.XPub)
+        , MkKeyFingerprint k Address
+        , SoftDerivation k
+        , Typeable n
+        , WalletKey k
+        )
+    => ctx
+    -> WalletId
+    -> k 'AccountK XPub
+    -> Cosigner
+    -> CredentialType
+    -> ExceptT ErrAddCosignerKey IO ()
+updateCosigner ctx wid accXPub cosigner cred = db & \DBLayer{..} -> do
+    mapExceptT atomically $ do
+        cp <- withExceptT ErrAddCosignerKeyNoSuchWallet $ withNoSuchWallet wid $
+              readCheckpoint (PrimaryKey wid)
+        let s = mapLeft toHandleErr $
+                addCosignerAccXPub accXPub cosigner cred (getState cp)
+        case s of
+            Left err -> throwE err
+            Right s' ->
+                withExceptT ErrAddCosignerKeyNoSuchWallet $
+                    putCheckpoint (PrimaryKey wid) (updateState s' cp)
+  where
+    db = ctx ^. dbLayer @s @k
+    toHandleErr = \case
+        ActiveWallet -> ErrAddCosignerKeyActiveWallet
+        NoSuchCosigner _ _ -> ErrAddCosignerKeyNoSuchCosigner cosigner cred
+        AlreadyPresentKey _ -> ErrAddCosignerKeyAlreadyPresentKey cred
+        NoDelegationTemplate -> ErrAddCosignerKeyNoDelegationTemplate
```

Comment:
This works. It's a bit different from the way we've been defining the error hierarchy within the wallet in other places. Take for example the `ErrSelectAssets` data-type which wraps errors from the coin-selection module:

```hs
data ErrSelectAssets
    = ErrSelectAssetsCriteriaError ErrSelectionCriteria
     | ErrSelectAssetsSelectionError SelectionError
     | ErrSelectAssetsNoSuchWallet ErrNoSuchWallet
```

It wraps errors from the coin-selection module that are themselves sum-types; This way, we don't have to re-create new branches for every branch in child errors. So here, I'd rather opt for:

```hs
data ErrAddCosignerKey
    = ErrAddCosignerKeyNoSuchWallet ErrNoSuchWallet
     | ErrAddCosignerKeyAddCosigner ErrAddCosigner
```

which makes it very easy to bubble up.

Reply:
realized by @rvl commits